src/Microsoft.DotNet.XHarness.CLI/Commands/WASI/Engine/WasiTestCommand.cs

```diff
@@ -0,0 +1,132 @@
+        }
+      
+        logger.LogInformation($"Using wasm engine {Arguments.Engine.Value} from path {engineBinary}");
+        await PrintVersionAsync(Arguments.Engine.Value.Value, engineBinary);
+
+        var cts = new CancellationTokenSource();
+        try
+        {
+            var engineArgs = new List<string>();
+            engineArgs.Add(Arguments.SubCommand);
+            engineArgs.Add("--dir");
+            engineArgs.Add(".");
+            engineArgs.Add(Arguments.WasmFile);
+            engineArgs.AddRange(Arguments.EngineArgs.Value);
+            engineArgs.AddRange(PassThroughArguments);
+
+            var xmlResultsFilePath = Path.Combine(Arguments.OutputDirectory, "testResults.xml");
+            File.Delete(xmlResultsFilePath);
+
+            var stdoutFilePath = Path.Combine(Arguments.OutputDirectory, "wasi-console.log");
+            File.Delete(stdoutFilePath);
+
+            var processTask = processManager.ExecuteCommandAsync(
+                engineBinary,
+                engineArgs,
+                log: new CallbackLog(m => logger.LogInformation(m)),
+                Arguments.Timeout);
+
+            var tasks = new Task[]
+            {
+                processTask,
+                Task.Delay(Arguments.Timeout)
+            };
+
+            var task = await Task.WhenAny(tasks).ConfigureAwait(false);
+            if (task == tasks[^1] || cts.IsCancellationRequested || task.IsCanceled)
+            {
+                logger.LogError($"Tests timed out after {((TimeSpan)Arguments.Timeout).TotalSeconds}secs");
+                if (!cts.IsCancellationRequested)
+                    cts.Cancel();
+
+                return ExitCode.TIMED_OUT;
+            }
+
+            if (task.IsFaulted)
+            {
+                logger.LogError($"task faulted {task.Exception}");
+                throw task.Exception!;
+            }
+
+            var result = await processTask;
+
+            if (result.ExitCode != Arguments.ExpectedExitCode)
+            {
+                logger.LogError($"Application has finished with exit code {result.ExitCode} but {Arguments.ExpectedExitCode} was expected");
+                return ExitCode.GENERAL_FAILURE;
+            }
+            else
+            {
+                logger.LogInformation("Application has finished with exit code: " + result.ExitCode);
+                return ExitCode.SUCCESS;
+            }
+        }
+        catch (Win32Exception e) when (e.NativeErrorCode == 2)
+        {
+            logger.LogCritical($"The engine binary `{engineBinary}` was not found");
+            return ExitCode.APP_LAUNCH_FAILURE;
+        }
+        finally
+        {
+            if (!cts.IsCancellationRequested)
+            {
+                cts.Cancel();
+            }
+        }
+
+        Task PrintVersionAsync(WasmEngine engine, string engineBinary)
+        {
+            return Task.CompletedTask;
+        }
```

Comment:
The implementation from wasm test command can be used here too, just with `--version` as the argument.